### DISCUSSION TITLE
Using aries-opt engine

### DISCUSSION THREAD

**Discussion by jdavidrcamacho** [aiplan4eu/unified-planning#571] [Q&A] [answered]
Hello everyone,

I have a quick doubt about the engines used in unified-planning. 
I know we can use the `aries` engine, but can we use the `aries-opt` engine (mentioned [here](https://github.com/plaans/aries/tree/master/planning/unified/plugin#aries-solver))?

**Comment by arbimo** [comment, marked as answer]
Hi, aries-opt is not available by default because:

1. it is generally useless: there is very few cases where it would provide an optimality proof (specifically scheduling and non-recursive HTN). In all other cases, the solver would just run forever.
2. it would be misleading as the current problem kinds do not allow capturing these, specifically, missing the distinction between recursive and non recursive HTN.

If you are interested in plan quality, then you should use the `AnytimePlanning` operation mode to get a stream of solution of increasing quality. Just as aries-opt, it would stop with `SOLVED_OPTIMALLY` when the optimality of the last solution is proven.

If you really only want the **provably** optimal plan, you can register the `up_aries.AriesOpt` class manually: https://unified-planning.readthedocs.io/en/latest/notebooks/tutorial/planner-integration.html#Registering-the-engine

**Reply by jdavidrcamacho** [reply]
Thank you by the answer! I manage to register aries-opt and use it but my problem still remains maybe you can shed some light on why the reason might be or how to solve it. 

This happens for both aries and aries-opt and sometimes when a plan is generated it gives a slightly different answer. For example while I would expect the optimal solution for my problem to be and most of the time is:

```
aries Hiearchical SequentialPlan:
    switch_on(instrument0, satellite0)
    nop
    turn_to(satellite0, groundstation2, phenomenon6)
    calibrate(satellite0, instrument0, groundstation2)
    turn_to(satellite0, phenomenon4, groundstation2)
    take_image(satellite0, phenomenon4, instrument0, thermograph0)
    nop
    turn_to(satellite0, star5, phenomenon4)
    take_image(satellite0, star5, instrument0, thermograph0)
    nop
    turn_to(satellite0, phenomenon6, star5)
    take_image(satellite0, phenomenon6, instrument0, thermograph0)
```

Sometimes it generates this 

```
aries Hiearchical SequentialPlan:
    switch_on(instrument0, satellite0)
    nop
    turn_to(satellite0, groundstation2, phenomenon6)
    calibrate(satellite0, instrument0, groundstation2)
    turn_to(satellite0, phenomenon4, groundstation2)
    take_image(satellite0, phenomenon4, instrument0, thermograph0)
    switch_off(instrument0, satellite0)
    switch_on(instrument0, satellite0)
    nop
    turn_to(satellite0, groundstation2, phenomenon4)
    calibrate(satellite0, instrument0, groundstation2)
    turn_to(satellite0, star5, groundstation2)
    take_image(satellite0, star5, instrument0, thermograph0)
    nop
    turn_to(satellite0, phenomenon6, star5)
    take_image(satellite0, phenomenon6, instrument0, thermograph0)
```

Where it has steps that are unnecessary steps, i.e. switching a satellite on and off for no apparent reason. Do you know how can I make aries results to be deterministic and truly optimal?

**Comment by arbimo** [comment]
The solver will run a handful of strategies in parallel that would share solutions and no-good clauses, hence the search is not expected to be deterministic.

What is the metric you specified for the problem ? If there is no metric, the solver may return any plan. If you specifically require length-optimal plan aries-opt shouldn't and I would be interested in seeing the problem to analyze it.

**Reply by jdavidrcamacho** [reply]
I have tried 3 different ways:

`OneshotPlanner(name="aries", optimality_guarantee=PlanGenerationResultStatus.SOLVED_OPTIMALLY,)`, 

`OneshotPlanner(name="aries", optimality_guarantee=PlanGenerationResultStatus.SOLVED_SATISFICING,)`, and 

`AnytimePlanner(name="aries", anytime_guarantee="INCREASING_QUALITY")`  

where I also used `problem.add_quality_metric(up.model.metrics.MinimizeSequentialPlanLength())`

but still the weird plan show up from time to time (~ every 10 times I run it, 9 succeeded but 1 has those unnecessary steps)

**Comment by arbimo** [comment]
Well, you always use aries in the above examples. Try one of the following.

    solver = OneshotPlanner(name="aries-opt")
    # or directly
    sovler = up_aries.AriesOpt()

As you specify the name "aries", this is the planner that would be selected. Also your `optimality_guarantee` parameter has the wrong type.

**Reply by jdavidrcamacho** [reply]
I didn't mentioned but already tried  `OneshotPlanner(name="aries-opt")` and after a few trials it gave the same wrong results

**Reply by arbimo** [reply]
Then, could please send what's necessary to reproduce the problem.

**Reply by jdavidrcamacho** [reply]
Can you please send me an email (joao.camacho@visionspace.com)? I'll reply to it with the data and scripts

**Comment by jdavidrcamacho** [comment]
Closing it as we discussed the issue via email

**Reply by arbimo** [reply]
FTR, the problem was that to have an optimal plan, both an optimal planner and a metric is needed.
